src/Http/Headers/src/ContentDispositionHeaderValue.cs

```diff
@@ -675,11 +657,45 @@ private static string Encode5987(StringSegment input)
                                    '0', '1', '2', '3', '4', '5', '6', '7',
                                    '8', '9', 'A', 'B', 'C', 'D', 'E', 'F' };
 
-    private static void HexEscape(StringBuilder builder, char c)
+    private static void EncodeToUtf8Hex(Rune rune, StringBuilder builder)
     {
-        builder.Append('%');
-        builder.Append(HexUpperChars[(c & 0xf0) >> 4]);
-        builder.Append(HexUpperChars[c & 0xf]);
+        // Inspired by https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Text/Rune.cs TryEncodeToUtf8
```

Comment:
Not a fan of copying this code. How much would we lose if we just `stackalloc byte[4]` and called `TryEncodeToUtf8` and looped over the result to write to the `StringBuilder`?

Reply:
https://github.com/dotnet/aspnetcore/pull/55039#discussion_r1564112587

Reply:
Personally, I can live with the copy since it's an implementation of spec'd behavior.  I don't think we'll get out of sync in a way that reduces correctness, but we could well fail to benefit from future perf wins.  If there's a way around copying, that would be preferable, but I wouldn't block the PR for this.  Having said that, I also wouldn't go ahead without @BrennanConroy's approval.

Reply:
That's fair 👍